### DISCUSSION TITLE
DXF Rotate

### DISCUSSION THREAD

**Discussion by softy2k** [mozman/ezdxf#853] [General]
Hello

I am trying to rotate (mirror) a DXF file (gerber). Just rotate the whole DXF by a certain number of degrees, nothing special.
I state from the beginning that this is my first contact with Python and with DXF files.

So, I installed Python : `pip install ezdxf`, then `pip install ezdxf[draw]` (I don't know if the second step was necessary).

Then, I tried various options, but without success.
E.g:
```
import ezdxf
doc = ezdxf.readfile('c:/input.dxf')
msp = doc.modelspace()
angle = 0.7853981634 # 45 degrees in radians
msp.rotate(angle)
doc.saveas('c:/output.dxf')
```
, but I get an error message: `AttributeError: 'Modelspace' object has no attribute 'rotation'`

I also attach the file that I want to rotate.
Basically, I'm looking for a simple method (regardless of which) to rotate a DXF file.
Thanks in advance for any help.

[input.zip](https://github.com/mozman/ezdxf/files/11025961/input.zip)

**Comment by mozman** [comment]
You cannot rotate the modelspace, you have to rotate the entities of the modelspace:

```Python
for e in msp:
    e.rotate_z(angle)
```

This rotates all entities of the modelspace around the z-axis in the origin (0, 0), this may raise a `NotImplementedError` for unsupported entities.

BTW: Rotation is nothing special for a CAD application, but `ezdxf` is a file format interface and not a CAD application!

**Comment by softy2k** [comment]
Thank you very much, it works.
But I can't turn it so that what is up is down and vice versa. I also tried the x, y and z axis with various angles (-180 to 180).
I must be doing something wrong, but I still don't know what...

**Reply by mozman** [reply]
Mirroring is scaling by -1!

**Comment by softy2k** [comment]
I've tried various options, but still haven't succeeded.
```
for e in msp:
    e.scale(-1, 1, 1)
```

If I try: 
```
for e in msp:
    e.scale(-1)
```
`TypeError: DXFGraphic.scale() missing 2 required positional arguments: 'sy' and 'sz'`

**Comment by mozman** [comment]
```Python
import ezdxf

doc = ezdxf.new()
msp = doc.modelspace()
pline = msp.add_lwpolyline(([(1, 1), (3, 2), (1, 3)]))
doc.saveas("test.dxf")

pline.scale(-1, 1, 1)
doc.saveas("test_sx.dxf")
````

test.dxf:
![image](https://user-images.githubusercontent.com/4230966/226678461-3a288e08-188d-405a-920f-db370d173bcc.png)

test_sx.dxf:
![image](https://user-images.githubusercontent.com/4230966/226678556-a66c6cfe-25fc-47e1-a8d6-332701f41015.png)

**Comment by softy2k** [comment]
Thanks for the reply.

I found the problem. And the code I wrote was OK, but the problem was the application with which I opened the DXFs.

More precisely with
```
DWG TrueView 2023
DWG FastView
DoubleCAD XT 5
and some online web viewer
```

DXFs were displayed correctly rotated.

With the applications:
```
FreeCAD 0.20
DXF Viewer
LibreCAD
```
the same rotated DXFs were not displayed correctly.
For this reason I was testing and could not find the solution.

I don't know why this happens, I used the applications with the default settings, on Windows 10 Pro (64 bits)
